### ISSUE TITLE
[v4] RenderTexture::saveToFile is leaking large chunks of memory

### ISSUE THREAD

**Issue by rh101** [cocos2d/cocos2d-x#20408]
- cocos2d-x version: 4.0.0
- devices test on: Windows 10
- developing environments
   - NDK version: 
   - Xcode version:
   - VS version: 2019
   - browser type and version:

Steps to Reproduce:

1. Run the cpp-tests. 
2. Go to test 38 "Node: RenderTexture"
3. Repeatedly click any of the "Save Image" options.

Memory usage will keep rising, perhaps as much as the size of the texture that is being saved out.

This issue is not happening with v3.17.2 or the current v3 branch from github.

This happens regardless of which format you choose to save the file as (PNG and JPG).

**Comment by rh101** [comment]
In the image below you can see the memory usage going up every time `RenderTexture::saveToFile()` is called.

![image](https://user-images.githubusercontent.com/8603230/70586441-2e9da700-1c1b-11ea-9460-23ac5454d467.png)

![image](https://user-images.githubusercontent.com/8603230/70586384-0615ad00-1c1b-11ea-901a-e3f36767b3c4.png)

![image](https://user-images.githubusercontent.com/8603230/70586426-23e31200-1c1b-11ea-966f-f925d2e2754d.png)

**Comment by rh101** [comment]
```
void Texture2DGL::getBytes(std::size_t x, std::size_t y, std::size_t width, std::size_t height, bool flipImage, std::function<void(const unsigned char*, std::size_t, std::size_t)> callback)
{
    GLint defaultFBO = 0;
    glGetIntegerv(GL_FRAMEBUFFER_BINDING, &defaultFBO);

    GLuint frameBuffer = 0;
    glGenFramebuffers(1, &frameBuffer);
    glBindFramebuffer(GL_FRAMEBUFFER, frameBuffer);
    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, _textureInfo.texture, 0);
    glPixelStorei(GL_PACK_ALIGNMENT, 1);

    auto bytePerRow = width * _bitsPerElement / 8;
    unsigned char* image = new unsigned char[bytePerRow * height];
    glReadPixels(x,y,width, height,GL_RGBA,GL_UNSIGNED_BYTE, image);

    if(flipImage)
    {
        unsigned char* flippedImage = new unsigned char[bytePerRow * height];
        for (int i = 0; i < height; ++i)
        {
            memcpy(&flippedImage[i * bytePerRow],
                   &image[(height - i - 1) * bytePerRow],
                   bytePerRow);
        }
        callback(flippedImage, width, height);
        CC_SAFE_DELETE_ARRAY(flippedImage);
    } else
    {
        callback(image, width, height);
        CC_SAFE_DELETE_ARRAY(image);
    }

    glBindFramebuffer(GL_FRAMEBUFFER, defaultFBO);
    glDeleteFramebuffers(1, &frameBuffer);
}
```

If flipImage is true, then it allocates memory for the flipped image, and also frees it after it's used, but it never frees the original image data that was also allocated.

It needs to be:
```

    if(flipImage)
    {
        unsigned char* flippedImage = new unsigned char[bytePerRow * height];
        for (int i = 0; i < height; ++i)
        {
            memcpy(&flippedImage[i * bytePerRow],
                   &image[(height - i - 1) * bytePerRow],
                   bytePerRow);
        }
        CC_SAFE_DELETE_ARRAY(image); // This fixes the leak

        callback(flippedImage, width, height);
        CC_SAFE_DELETE_ARRAY(flippedImage);
    } else
```

**Comment by Mee-gu** [comment]
fixed in [PR 20409](https://github.com/cocos2d/cocos2d-x/pull/20409)

**Comment by rh101** [comment]
Out of curiosity, wouldn't it be better to free it before the callback, in case the user does more processing on the image data in memory, and perhaps even allocates more memory from the image in the callback.  At the moment there would be this chunk of memory being used by the data in `image` that serves no purpose.

If it's freed before, then at least that reduces the impact on the current memory usage before the callback occurs, since the data in `image` is not used at all after it's copied to `flippedImage`.

```
        CC_SAFE_DELETE_ARRAY(image);
        callback(flippedImage, width, height);
        CC_SAFE_DELETE_ARRAY(flippedImage);
```

**Comment by Mee-gu** [comment]
yep, it's better to release the unused image before the callback . I will fix it later.

**Comment by qoozta** [comment]
How about the TextureCubeGL::getBytes? I think it's using the same structure.
Also in TextureMTL.mm may have the same leaks.

**Comment by rh101** [comment]
> How about the TextureCubeGL::getBytes? I think it's using the same structure.
> Also in TextureMTL.mm may have the same leaks.

From the look of it that doesn't seem to be the case, so have you actually checked it yourself? Is it leaking memory? 

In `Utils::getTextureBytes` I can see the following lines:
```
auto bytePerRow = rectWidth * getBitsPerElement(texture.pixelFormat) / 8;
unsigned char* image = new (std::nothrow) unsigned char[bytePerRow * rectHeight];
if(image != nullptr)
{
    [copiedTexture getBytes:image bytesPerRow:bytePerRow fromRegion:imageRegion mipmapLevel:0];
    swizzleImage(image, rectWidth, rectHeight, texture.pixelFormat);
}
callback(image, rectWidth, rectHeight);
CC_SAFE_DELETE_ARRAY(image);
[copiedTexture release];
```

The `CC_SAFE_DELETE_ARRAY(image);` is deleting the allocated memory, so what exactly are you referring to?

**Comment by qoozta** [comment]
> > How about the TextureCubeGL::getBytes? I think it's using the same structure.
> > Also in TextureMTL.mm may have the same leaks.
> 
> From the look of it that doesn't seem to be the case, so have you actually checked it yourself? Is it leaking memory?
> 
> In `Utils::getTextureBytes` I can see the following lines:
> 
> ```
> auto bytePerRow = rectWidth * getBitsPerElement(texture.pixelFormat) / 8;
> unsigned char* image = new (std::nothrow) unsigned char[bytePerRow * rectHeight];
> if(image != nullptr)
> {
>     [copiedTexture getBytes:image bytesPerRow:bytePerRow fromRegion:imageRegion mipmapLevel:0];
>     swizzleImage(image, rectWidth, rectHeight, texture.pixelFormat);
> }
> callback(image, rectWidth, rectHeight);
> CC_SAFE_DELETE_ARRAY(image);
> [copiedTexture release];
> ```
> 
> The `CC_SAFE_DELETE_ARRAY(image);` is deleting the allocated memory, so what exactly are you referring to?

In TextureMTL::getBytes, isn't the same situation?
```
if(!flipImage)
        {
            unsigned char* flippedImage = new unsigned char[bytePerRow * height];
            for (int i = 0; i < height; ++i)
            {
                memcpy(&flippedImage[i * bytePerRow],
                       &image[(height - i - 1) * bytePerRow],
                       bytePerRow);
            }
            callback(flippedImage, width, height);
            CC_SAFE_DELETE_ARRAY(flippedImage);
        }
```

**Comment by rh101** [comment]
> In TextureMTL::getBytes, isn't the same situation?

If you look at the code you posted, you will see that it is freeing the data, so again, are you asking for the sake of knowing, or have you actually encountered a memory leak? 

If you have experienced some kind of memory leak in that code, then debug it, and verify the source of the memory leak, and then create a new issue for it.  The problem detailed in this issue has been addressed.

**Comment by rh101** [comment]
This issue was fixed in #20409
